File: 2943.maximize-area-of-square-hole-in-grid/solution.go
func maximizeSquareHoleArea(n int, m int, hBars []int, vBars []int) int {
    gap := min(maxGap(n, hBars), maxGap(m, vBars)) + 1
    return gap * gap
}

func maxGap(l int, bars []int) int {
    slices.Sort(bars)
    r, m := 1, 1
    for i := 1; i < len(bars); i++ {
        if bars[i] == bars[i-1] + 1 {
            r++
        } else {
            r = 1
        }
        if r > m {
            m = r
        }
    }
    return m
}

File: 2943.maximize-area-of-square-hole-in-grid/solution.kt
class Solution {
    fun maximizeSquareHoleArea(n: Int, m: Int, hBars: IntArray, vBars: IntArray): Int {
        val gap = minOf(maxGap(hBars), maxGap(vBars)) + 1
        return gap * gap        
    }

    fun maxGap(bars: IntArray): Int {
        return if (bars.size == 1) {
            1
        } else {
            var run = 1
            bars.sorted()
                .zipWithNext()
                .maxOf { (a, b) ->
                    if (a + 1 == b) {
                        run++
                    } else {
                        run = 1
                    }
                    run
                }
        }
    }
}